### PR TITLE
Add Cache Purge Command

### PR DESCRIPTION
**Pull request by luiztiago**
## Description

Customers could use the CLI to purge cache as a similar command used on WP sites.

## Steps to Test

1. Check out PR.
1. Run `npm run build` and `npm link`
1. Run `node ./dist/bin/vip @4716 cache purge-url https://luiztiago-test-node.go-vip.net/`
1. Verify `Purged URLs` success message

### CODE REVIEW

File: __tests__/bin/vip-cache-purge.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		} );
+
+		await cachePurgeCommand( args, opts );
+
+		expect( console.log ).toHaveBeenCalledTimes( 1 );
+		expect( console.log ).toHaveBeenCalledWith(
+			/* eslint-disable indent */
+	'- Purged URLs: url'
+				/* eslint-enable indent */
+		);
```
Comment (chriszarate) [inline __tests__/bin/vip-cache-purge.js]:
```suggestion
		expect( console.log ).toHaveBeenCalledWith( '- Purged URLs: url' );
```

File: npm-shrinkwrap.json [hunk 2]
```diff
@@ -10146,6 +10146,14 @@
           "resolved": "https://registry.npmjs.org/ansi-regex/-/ansi-regex-3.0.1.tgz",
           "integrity": "sha512-+O9Jct8wf++lXxxFc4hc8LsjaSq0HFzzL7cVsw8pRDIPdjKD2mT4ytDZlLuSBZ4cLKZFXIrMGO7DbQCtMJJMKw=="
         },
+        "cli-table": {
```
Comment (chriszarate) [inline npm-shrinkwrap.json]: hmm why did this file change?

File: src/bin/vip-cache-purge.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * Internal dependencies
+ */
+import { appQuery, purgeCache } from 'lib/api/cache-purge';
+import command from 'lib/cli/command';
+import { trackEvent } from 'lib/tracker';
+import * as exit from 'lib/cli/exit';
+
+const examples = [
+	{
+		usage: 'vip @123.production cache purge <URL1> <URL2>',
```
Comment (chriszarate) [inline src/bin/vip-cache-purge.js]: Is this how we want to support purging multiple URLs? Or would it be better to allow passing in multiple URLs via `stdin` or a text file?

File: src/bin/vip-cache-purge.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+		description: 'Clear cache for a Node.js URL',
```
Comment (chriszarate) [inline src/bin/vip-cache-purge.js]: This works for WP sites too, no?

File: src/bin/vip-cache-purge.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+	await trackEvent( 'cache_purge_command_execute', trackingParams );
+
+	let purgeCacheObject = {};
+	try {
+		purgeCacheObject = await purgeCache( opt.app.id, opt.env.id, urls );
+	} catch ( err ) {
+		await trackEvent( 'cache_purge_command_error', { ...trackingParams, error: err.message } );
+
+		exit.withError( `Failed to purge cache object error: ${ err.message }` );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge.js]:
```suggestion
		exit.withError( `Failed to purge cache object: ${ err.message }` );
```

File: src/bin/vip-cache-purge.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	}
+
+	await trackEvent( 'cache_purge_command_success', trackingParams );
+
+	const output = [
+		`- Purged URLs: ${ purgeCacheObject.urls.join( ',' ) }`,
+	];
+
+	console.log( output.join( '\n' ) );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge.js]: this will make for much better output in the terminal

```suggestion
	purgeCacheObject.urls.forEach( ( url ) => {
		console.log( `- Purged URL: ${ url }` );
	} );
```

File: src/bin/vip-cache-purge.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+command( {
+	appContext: true,
+	appQuery,
+	envContext: true,
+	requiredArgs: 1,
+	wildcardCommand: true,
```
Comment (chriszarate) [inline src/bin/vip-cache-purge.js]: i forget what this means. does it allow the argument to take any form?

File: src/bin/vip-cache-purge.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		app_id: opt.app.id,
+		command: 'vip cache purge',
+		env_id: opt.env.id,
+	};
+
+	await trackEvent( 'cache_purge_command_execute', trackingParams );
+
+	let purgeCacheObject = {};
+	try {
+		purgeCacheObject = await purgeCache( opt.app.id, opt.env.id, urls );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge.js]: should we do any inspection of the input?

File: src/lib/api/cache-purge.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+		appId,
+		envId,
+		urls,
+	};
+
+	const response = await api.mutate( { mutation, variables } );
+
+	const { data: { purgeCacheObject } } = response;
+	if ( ! purgeCacheObject ) {
+		throw new Error( 'Something went wrong :(' );
```
Comment (chriszarate) [inline src/lib/api/cache-purge.js]: Can we be more specific here? Will there be an error returned? Is this even necessary or can we handle the error in the calling code?

File: src/bin/vip-cache-purge.js (hunk 7 above)
Comment (luiztiago) [reply src/bin/vip-cache-purge.js]: @chriszarate  This is used to accept args when using multiple URLs

File: src/bin/vip-cache-purge.js (hunk 3 above)
Comment (luiztiago) [reply src/bin/vip-cache-purge.js]: @chriszarate Do you have an example with stdin or text file working? I've tried a few things but with no success

Comment (chriszarate) [reply src/bin/vip-cache-purge.js]: here's an example using a file (via an option):
https://github.com/Automattic/vip/blob/5372034e86d3aba96358286d6eef01a5c879432f/src/bin/vip-config-envvar-set.js#L67

this is all that's necessary, since someone can use this option to pass in `stdin`, e.g.:

`echo "my value" | vip @4092 config envvar set MYVAR --from-file=/dev/stdin`

File: npm-shrinkwrap.json (hunk 2 above)
Comment (ingeniumed) [reply npm-shrinkwrap.json]: For me, this and the colours bit below get removed when I run `npm install` so not sure why this got added here. I repeated that 2-3 times and it got removed for me

File: src/bin/vip-cache-purge.js [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+ */
+import { appQuery, purgeCache } from 'lib/api/cache-purge';
+import command from 'lib/cli/command';
+import { trackEvent } from 'lib/tracker';
+import { readFromFile } from 'lib/read-file';
+import * as exit from 'lib/cli/exit';
+
+const examples = [
+	{
+		usage: 'vip @123.production cache purge <URL>',
```
Comment (ingeniumed) [inline src/bin/vip-cache-purge.js]: Should this be URLs, based on the design session and the list below in the command?

File: src/bin/vip-cache-purge.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+		const value = await readFromFile( opt.fromFile );
+		if ( value ) {
+			urls = value.split( '\n' ).map( url => url.trim() );
+		}
+	}
+
+	if ( ! urls.length ) {
+		await trackEvent( 'cache_purge_command_error', trackingParams );
+
+		exit.withError( 'You need at least an URL to purge cache' );
```
Comment (ingeniumed) [inline src/bin/vip-cache-purge.js]: Could I suggest changing this to `You need at least one URL to purge the cache`?

File: src/lib/envvar/read-file.js [hunk 12]
```diff
@@ -6,25 +6,16 @@
...
 /**
  * Internal dependencies
  */
-import * as exit from 'lib/cli/exit';
 import { debug } from 'lib/envvar/logging';
+import { readFromFile } from '../read-file';
 
 export async function readVariableFromFile( path ): Promise<string> {
 	debug( `Loading variable value from file ${ JSON.stringify( path ) }` );
+	const data = await readFromFile( path );
```
Comment (ingeniumed) [inline src/lib/envvar/read-file.js]: Could return it inline here

File: src/bin/vip-cache-purge.js (hunk 10 above)
Comment (chriszarate) [reply src/bin/vip-cache-purge.js]: I'm open to discussion, but I think I would prefer a single entrypoint to purging multiple URLs. The `--from-file` option makes the most sense to me. If I'm bringing a list of URLs to this command, it makes sense that I'd have them in a newline-separated list instead of a space-separated list.

So I'd argue for keeping this usage the way it is and only allowing one URL:

```
vip @123.production cache purge <URL>
```

For multiple URLs, users can use `--from-file` or just run the command multiple times:

```
vip @123.production cache purge <URL1>
vip @123.production cache purge <URL2>
```

Comment (chriszarate) [reply src/bin/vip-cache-purge.js]: I think I'd also advocate for renaming the command to `purge-url`. I can envision adding other commands in the future; perhaps purging via a `BAN` directive (pattern) or potentially `purge-all` or purging other types other types of cache. So being specific here will help us avoid ambiguity in the future.

Comment (luiztiago) [reply src/bin/vip-cache-purge.js]: @chriszarate I agree, I'm renaming it and removing the list of URLs using args.

File: src/lib/api/cache-purge.js (hunk 9 above)
Comment (luiztiago) [reply src/lib/api/cache-purge.js]: @chriszarate If the command throws an error, it will be printed on the CLI command, as the domain exception. This error message is just a "backup" if the purgeCacheObject is null and the mutation is not throwing an error.

Comment (chriszarate) [reply src/lib/api/cache-purge.js]: Understood, it's just not particularly useful as an error message. If you removed this check/throw, it would raise an exception whenever you tried to destructure from `null`, which is probably just as useful, if not more so.

If there are situations where we think we might return `null`, we should probably convert them to useful error messages in the API.

Comment (luiztiago) [reply src/lib/api/cache-purge.js]: @chriszarate I agree 👍

File: src/bin/vip-cache-purge-url.js [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+ */
+import { appQuery, purgeCache } from 'lib/api/cache-purge';
+import command from 'lib/cli/command';
+import { trackEvent } from 'lib/tracker';
+import { readFromFile } from 'lib/read-file';
+import * as exit from 'lib/cli/exit';
+
+const examples = [
+	{
+		usage: 'vip @123.production cache purge-url <URL>',
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]: We generally don't add the `@` shorthand to usage commands since it is provided separately by the echoed options help text

```suggestion
		usage: 'vip cache purge-url <URL>',
```

File: src/bin/vip-cache-purge-url.js [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+		description: 'Purge cache for a URL',
+	},
+	{
+		usage: 'vip @123.production cache purge-url --from-file=/dev/vip/urls.txt',
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
		usage: 'vip cache purge-url --from-file=/dev/vip/urls.txt',
```

File: src/bin/vip-cache-purge-url.js [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+		console.log( `- Purged URL: ${ url }` );
+	} );
+}
+
+command( {
+	appContext: true,
+	appQuery,
+	envContext: true,
+	wildcardCommand: true,
+	usage: 'vip @123.production cache purge-url <URL>',
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
	usage: 'vip cache purge-url <URL>',
```

File: src/bin/vip-cache-purge-url.js [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+} )
+	.option( 'from-file', 'Read URLs from file (useful for multiline input)' )
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
	.option( 'from-file', 'Read URLs from file (useful to purge multiple URLs)' )
```

File: src/bin/vip-cache.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ * Internal dependencies
+ */
+import command from 'lib/cli/command';
+
+command( {
+	requiredArgs: 2,
+} )
+	.command( 'purge-url', 'Purge cache' )
```
Comment (chriszarate) [inline src/bin/vip-cache.js]:
```suggestion
	.command( 'purge-url', 'Purge page cache' )
```

File: src/bin/vip-cache-purge-url.js [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+import { appQuery, purgeCache } from 'lib/api/cache-purge';
+import command from 'lib/cli/command';
+import { trackEvent } from 'lib/tracker';
+import { readFromFile } from 'lib/read-file';
+import * as exit from 'lib/cli/exit';
+
+const examples = [
+	{
+		usage: 'vip cache purge-url <URL>',
+		description: 'Purge cache for a URL',
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]: Today's session taught me that we should be very specific with our terms :)

```suggestion
		description: 'Purge a URL from page cache',
```

File: src/bin/vip-cache-purge-url.js [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+	},
+	{
+		usage: 'vip cache purge-url --from-file=/dev/vip/urls.txt',
+		description: 'Purge cache for multiple URLs',
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
		description: 'Purge multiple URLs from page cache',
```

File: src/bin/vip-cache-purge-url.js [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+export async function cachePurgeCommand( urls = [], opt = {} ): void {
+	const trackingParams = {
+		app_id: opt.app.id,
+		command: 'vip cache purge-url',
+		env_id: opt.env.id,
+		from_file: !! opt.fromFile,
+	};
+
+	await trackEvent( 'cache_purge_command_execute', trackingParams );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
	await trackEvent( 'cache_purge_url_command_execute', trackingParams );
```

File: src/bin/vip-cache-purge-url.js [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+
+	if ( opt.fromFile ) {
+		const value = await readFromFile( opt.fromFile );
+		if ( value ) {
+			urls = value.split( '\n' ).map( url => url.trim() );
+		}
+	}
+
+	if ( ! urls.length ) {
+		await trackEvent( 'cache_purge_command_error', trackingParams );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
		await trackEvent( 'cache_purge_url_command_error', { ...trackingParams, error: 'Missing URL' } );
```

File: src/bin/vip-cache-purge-url.js [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+		exit.withError( 'You need at least one URL to purge the cache' );
+	}
+
+	let purgeCacheObject = {};
+	try {
+		purgeCacheObject = await purgeCache( opt.app.id, opt.env.id, urls );
+	} catch ( err ) {
+		await trackEvent( 'cache_purge_command_error', { ...trackingParams, error: err.message } );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
		await trackEvent( 'cache_purge_url_command_error', { ...trackingParams, error: err.message } );
```

File: src/bin/vip-cache-purge-url.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+		const value = await readFromFile( opt.fromFile );
+		if ( value ) {
+			urls = value.split( '\n' ).map( url => url.trim() );
+		}
+	}
+
+	if ( ! urls.length ) {
+		await trackEvent( 'cache_purge_command_error', trackingParams );
+
+		exit.withError( 'You need at least one URL to purge the cache' );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
		exit.withError( 'Please supply at least one URL.' );
```

File: src/bin/vip-cache-purge-url.js [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	let purgeCacheObject = {};
+	try {
+		purgeCacheObject = await purgeCache( opt.app.id, opt.env.id, urls );
+	} catch ( err ) {
+		await trackEvent( 'cache_purge_command_error', { ...trackingParams, error: err.message } );
+
+		exit.withError( `Failed to purge cache object: ${ err.message }` );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
		exit.withError( `Failed to purge URL(s) from page cache: ${ err.message }` );
```

File: src/bin/vip-cache-purge-url.js [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+	}
+
+	await trackEvent( 'cache_purge_command_success', trackingParams );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
	await trackEvent( 'cache_purge_url_command_success', trackingParams );
```

File: src/bin/vip-cache-purge-url.js (hunk 18 above)
Comment (luiztiago) [reply src/bin/vip-cache-purge-url.js]: @chriszarate All suggestions are much better, thanks! :)

**Review by ingeniumed — APPROVED** [review]

File: npm-shrinkwrap.json [hunk 26]
```diff
@@ -10104,8 +10104,8 @@
       "dev": true
     },
     "lando": {
-      "version": "git+https://github.com/Automattic/lando-cli.git#d4d836c1ece783a2eb4aead589cf5d3889efa1bd",
```
Comment (chriszarate) [inline npm-shrinkwrap.json]: can you revert the changes to this file? should be unnecessary

File: src/bin/vip-cache-purge-url.js [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+
+	if ( opt.fromFile ) {
+		const value = await readFromFile( opt.fromFile );
+		if ( value ) {
+			urls = value.split( '\n' ).map( url => url.trim() );
+		}
+	}
+
+	if ( ! urls.length ) {
+		await trackEvent( 'cache_purge_url_command_error', trackingParams );
```
Comment (chriszarate) [inline src/bin/vip-cache-purge-url.js]:
```suggestion
		await trackEvent( 'cache_purge_url_command_error', { ...trackingParams, error: 'No URL provided' } );
```

**Review by chriszarate — CHANGES_REQUESTED** [review]

**Review by chriszarate — APPROVED** [review]
YES!!1!

**Comment by chriszarate** [comment]
Need to update tests, otherwise g2g

**Review by chriszarate — APPROVED** [review]
